### PR TITLE
Add feature to save `best_attack_token`

### LINKED ISSUE #12: Run test on attack with lowest loss
For some schedulers, like cosine, it possible for an attack at a prior step, and not the final step, to have a lower loss. We should keep track of the lowest train (or validation) loss attack and use that in the final step.

This is probably best done in conjunction with #11.

### PR DESCRIPTION
**Pull request by adarshan-intel**
Add feature to save `best_attack_token`

Fixes #12

### CODE REVIEW

**Review by mariusarvinte — COMMENTED** [review]
This is a good first step towards #12 but it's still missing the core functionality right?

The goal of #12 was to not only record the loss (a scalar), but also the exact attack tokens that cause that loss.

The final testing happens in https://github.com/IntelLabs/LLMart/blob/main/src/llmart/attack.py#L129 and consumes the `attack` variable returned by `train` in https://github.com/IntelLabs/LLMart/blob/main/src/llmart/attack.py#L120

Inside `train`, the `optimizer.step` modifies `attack` permanently at every step https://github.com/IntelLabs/LLMart/blob/main/src/llmart/attack.py#L301 (without keeping the `attack` that led to the lowest loss) - the goal of #12 would be to store the best `attack` if desired (via a new config e.g., `cfg.keep_best_attack`), and have `train` return that.

If it helps, this is a standard PyTorch flow, where `attack` is a subclass of `nn.Module` and it has a `param` attribute that actually stores the attack tokens - https://github.com/IntelLabs/LLMart/blob/main/src/llmart/model.py#L18

File: src/llmart/attack.py [hunk 1]
```diff
@@ -210,11 +210,16 @@ def train(
...
+    best_loss = float('inf')
+    best_attack_token = ""
+
     for step, inputs in (
         pbar := tqdm(iterable=enumerate(train_dl), total=len(train_dl), desc="steps")
     ):
         optimizer.zero_grad()
 
+        attack_param = attack.module.param if hasattr(attack, 'module') else attack.param
+        attack_token = tokenizer.decode(attack_param.argmax(dim=-1))
```
Comment (adarshan-intel) [inline src/llmart/attack.py:222]: When I run normal Python, I access attack.params. However, when using the Accelerate library, I need to access attack.module.params. This adjustment is necessary to handle both cases effectively

File: src/llmart/config.py [hunk 2]
```diff
@@ -448,8 +448,8 @@ class LLMartConf(CoreConf):
     early_stop: bool = True
     val_every: int = 50
     save_every: int = 50
-    max_new_tokens:int = 512
```
Comment (adarshan-intel) [inline src/llmart/config.py:451]: The recent change was made of `max_new_tokens` parameter, I added `keep_best_attack` var here. 
Please be careful when merging these changes.

File: src/llmart/attack.py [hunk 3]
```diff
@@ -327,6 +337,8 @@ def train(
         torch.save(accelerator.unwrap_model(attack).state_dict(), attack_path)
         log.info(f"{attack_path=}")
 
+    cfg.keep_best_attack = best_attack_token
```
Comment (mariusarvinte) [inline src/llmart/attack.py:340]: This should be `attack.param = best_attack_param`; that way the returned `attack` contains the tokens which cause the lowest loss

File: src/llmart/attack.py (hunk 1 above)
Comment (mariusarvinte) [inline src/llmart/attack.py:222]: I don't think we need the decoded tokens at all though, just the best param. The `test` function takes care of converting to strings later.

File: src/llmart/attack.py (hunk 3 above)
Comment (adarshan-intel) [reply src/llmart/attack.py:340]: I don't think we can store directly, first we need to encode to tensors right?

Comment (mariusarvinte) [reply src/llmart/attack.py:340]: You can just capture `best_attack_param = attack.param` directly (the tensor) and update it at the end if the flag is true; there's no need to decode in the first place, right?

**Review by mariusarvinte — APPROVED** [review]
This works as intended, when running
```
ts -nfG1 accelerate launch -m llmart model=deepseek-r1-distill-llama-8b data=basic attack.suffix=10 per_device_bs=64 val_every=5 keep_best_attack=1 steps=50
```

The best attack is at step 47
```
tokenizer.decode(best_attack_data.argmax(dim=-1)) = ' pornofilmUnexpected Joey waypoints Nou_functions("NEW inequalities APC'
```

And gets used at the final test in step 50
